docs/design/uplink-scopes.md

```diff
@@ -0,0 +1,67 @@
+# Title: Uplink Scopes
+
+## Abstract
+
+This design document proposes the data we need to store inside of a share in order to access files. Currently, the uplink configuration yaml file includes distinct entries for the API key, the satellite url, and the root encryption key. This proposal is to bundle an API key, a satellite url, and a list of revealed paths into a scope, much like a kubectl context.
+
+## Background
+
+In order to access objects, one must provide an API key to a given satellite. This API key is tied to a project which allows the satellite to look up the specific bucket being requested and do the things it needs to do. API keys are moving to be macaroons where they can be limited to only allow reads or writes, or expire, etc. Those features can easily be implemented and don't interact much the overall uplink command experience: you have access to do some actions or you don't.
+
+There are complications when it comes to restricting to a whitelist of buckets and/or path prefixes inside of those buckets. For example, if a macaroon is limited to only allow access to paths beneath `sj://bucket/a/b`, then it would be impossible to list the root. Additionally, the clients and servers all operate on **encrypted** paths, making it impossible for the server to issue the list and filter, because one would have to reveal the root encryption key for the bucket for the client to be able to decrypt the names.
+
+## Design
+
+Uplink will operate on a set of "scopes". Each scope contains an API key, a satellite url, an optional project key, and a list of bucket shares. In Go,
+
+```go
+type Scope struct {
+    APIKey               *storj.Key // or something else
+    SatelliteURL         string
+    ProjectEncryptionKey []byte
+    BucketShares         []BucketShare
+}
+```
+
+A bucket share contains an encrypted path, the matching unencrypted path, the key beginning at that path, and the bucket the path is contained in. In Go,
+
+```go
+type BucketShare struct {
+    Bucket          string
+    EncryptedPath   string
+    UnencryptedPath string
```

Comment:
Note that we have the `storj.Path` type (alias to `string`) that we use for paths within a bucket. It's good for both encrypted and unencrypted paths. Perhaps, use it here instead of `string`.

Reply:
https://github.com/storj/storj/blob/master/pkg/storj/path.go#L11